[Richard Hew Jia Liang] iP

ricketytoc opened the pull request:
# RichiePro
> "Your ind is for having ideas, not holding them." - David Allen
RichiePro frees your mind of having to remember things you need to do.It's, 

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is, 

1. download it from [here](www.google.com)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 👍 

and it is **FREE**!

Features:

- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders

Commit:
Renamed Chat bot from Duke to Richie and implemented an initial skeletal version of the chatbot (ie. greeting the user and exit)

Commit:
Allows chatbot to echo user commands and detects the "bye" command to exit

Commit:
added ability to store whatever text entered by the user and display them back to the user when required

Commit:
Added a task class

Commit:
Added the ability to mark tasks as done

Commit:
Added support for tracking three types of tasks : ToDos, Deadlines, Events

Commit:
Added in a set of inputs and the expected outputs (ie. added a test case)

Commit:
Allowed the bot to handle errors or incorrect inputs by adding a InvalidInputException class and a few try catch blocks

Commit:
Added support for deleting task from the list and also adding a try catch block for marking and deleting to handle errors

Commit:
Add javadocs

Commit:
Adds the ability to save the tasks in a text file whenever the task list changes.

Commit:
Handle case where data file does not exist

Commit:
Add data file and Richie Exception class

Commit:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Handle case where data file does not exist
  Adds the ability to save the tasks in a text file whenever the task list changes.
  Add javadocs

Commit:
Teach ChatBot how to understand date and time

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Teach ChatBot how to understand date and time

Commit:
organise the code into different classes (Storage, Parser, TaskList) and also organised the code into different packages

Commit:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  organise the code into different classes (Storage, Parser, TaskList) and also organised the code into different packages

Commit:
Restore some changes that were not in the version control.

Commit:
Fix programmme printing the bye message twice when exiting

Commit:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Fix programmme printing the bye message twice when exiting
  Restore some changes that were not in the version control.

Commit:
Removed extra files

Commit:
Automate project builds using Gradle

Commit:
Add Junit tests for Storage and Parser classes.

Commit:
Create JAR file

Commit:
Add javadoc to the classes in the duke and the duke.task packages

Commit:
Adhere to coding standard

Commit:
Find a task by searching for a keyword.

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add javadoc to the classes in the duke and the duke.task packages

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Find a task by searching for a keyword.

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Adhere to coding standard

Commit:
Add javadocs for new public methods associated with the find functionality

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.LocalDateTime;
+
```
Imports are not arranged in lexicographical order. Perhaps rearrange them using your IDE according to the CheckStyle rules when you complete the A-CheckStyle branch?

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+            String stringDateTime = input;
+            String[] dateTimeArray = stringDateTime.split(" ", 2);
+            String stringDate = dateTimeArray[0];
+            String stringTime = dateTimeArray[1];
```
Perhaps you could rename these variables? For example, rename `stringDateTime` to `dateTimeString` would make it easier to read.

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                try {
+                    if (!stringArray[1].contains("/by")) {
+                        throw new RichieException("OOPS!! please enter '/by' followed by a date and " +
+                                "time that the task should be done by");
```
'+' should be on a new line. (80:103) [OperatorWrap]
Perhaps you could also insert a line break such that the + operator is not needed?

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                    String[] stringArray3 = stringArray2[1].split(" /to ", 2);
+                    String stringFromDateTime = stringArray3[0];
+                    String stringToDateTime = stringArray3[1];
+                    return new EventCommand(stringArray2[0], convertInputDateAndTimeIntoLocalDateTime(stringFromDateTime),
```
Line exceeds hard limit of 120 characters. Perhaps you could insert a line break or shorten the method name?

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+        } else if (userInput.equals("list")) {
+            return new ListCommand();
+        } else {
+            String[] stringArray = userInput.split(" ", 2);
```
Could have a better name for `stringArray`. Perhaps use something like `argumentArray` or `parameterArray`.

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                                + "'/from' and '/to'");
+                    }
+                    String[] stringArray2 = stringArray[1].split(" /from ", 2);
+                    String[] stringArray3 = stringArray2[1].split(" /to ", 2);
```
Avoid using numbers or case to distinguish names. Perhaps you could rename them to `fromDateTime` and `toDateTime` instead?

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                    Integer.parseInt(stringTime.substring(2, 4)));
+            LocalDateTime dateTime = LocalDateTime.of(date, time);
+            return dateTime;
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new RichieException("Date and Time entered is in the wrong format!");
+        } catch (DateTimeException e) {
+            throw new RichieException("Date and Time entered is in the wrong format!");
```
These two catch blocks have the same body. Perhaps you could combine them or change one of them to have a different message?

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+                    Integer.parseInt(dayMonthYearArray[0]));
+            LocalTime time = LocalTime.of(Integer.parseInt(stringTime.substring(0, 2)),
+                    Integer.parseInt(stringTime.substring(2, 4)));
+            LocalDateTime dateTime = LocalDateTime.of(date, time);
```
The local variable `dateTime` is redundant as it is immediately returned. Perhaps you could inline the variable.
Example: `return LocalDateTime.of(date, time);`

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+            LocalDateTime dateTime = LocalDateTime.of(date, time);
+            return dateTime;
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new RichieException("Date and Time entered is in the wrong format!");
```
Avoid magic literals. Perhaps extract these strings to constants?

jiakai-17 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+                            + event.getFrom().toString() + "/"
+                            + event.getTo().toString());
+                    bufferedWriter.append("\n");
+                }
```
Perhaps you could include a final `else` clause to detect errors.

jiakai-17 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+ * deleting/adding tasks and marking tasks as done/undone
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
```
Perhaps rename this variable to not have the same name as the class.

jiakai-17 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,139 @@
+     * @return A Command object depending on what command was called by the user
+     * @throws RichieException Gives useful exception messages depending on the user input
+     */
+    public static Command parse(String userInput) throws RichieException {
```
This method is too long (>30 LoC). Perhaps you could shorten it to make it easier for others to understand?

jiakai-17 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class Ui {
+    private Scanner sc = new Scanner(System.in);
+    private String CHATBOT_NAME = "Richie";
```
Perhaps you could mark the constant `CHATBOT_NAME` as final?

jiakai-17 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * Prints the error message associated with loading data
+     */
+
```
Perhaps you could remove this blank line to align with the style guide?

jiakai-17 commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.time.LocalDateTime;
+
+public class DeadlineCommand extends Command {
```
Perhaps you could add Javadocs for public class, constructor and methods? It would be good to have them for ease of understanding.

jiakai-17 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+ * The parent class for all the commands in the Richie application
+ */
+public abstract class Command {
+    private boolean isExit;
```
The `isExit` variable does not seem to be used, perhaps it could be removed?

jiakai-17 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,21 @@
+        int length = taskList.getSize();
+        String result = "";
+        for (int i = 0; i < length; i++) {
+            result += (i + 1) + "." + taskList.getTask(i).toString() + "\n";
```
Perhaps you could use a StringBuilder when concatenating strings in a loop for better efficiency?

jiakai-17 commented on src/test/java/duke/RichieTest.java:
```diff
@@ -0,0 +1,17 @@
+
+class RichieTest {
+
+    @Test
```
Perhaps put some assertion statements in the Junit tests, instead of leaving them blank?

jiakai-17 reviewed (commented):
Overall good job, I can see the effort in ensuring good naming for variable and methods.

However, there is space for improvement in ensuring the package imports are correctly formatted using CheckStyle, which includes ordering and spacing.

Some of your methods are very long and complex to understand. Consider simplifying them.

There should be an empty line between description and parameter section for your Javadocs.

Also, do ensure that your files end with a newline.

jiakai-17 commented on src/test/java/duke/RichieTest.java:
```diff
@@ -0,0 +1,17 @@
+class RichieTest {
+
+    @Test
+    void run() {
```
The method for tests should be in the format: `featureUnderTest_testScenario_expectedBehavior()`

Commit:
Update build.gradle file

Commit:
Add a GUI to the chat bot

jianyangg commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+            }
+            return task;
+        } else {
+            return null;
```
Instead of returning `null`, could throwing an exception help handle this situation more gracefully?

jianyangg commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+     * Converts the LocalDateTime by into a String
+     * @return String that represents the by
+     */
+    public String convertByToString() {
```
A better naming for this function might help, since it may not be immediately apparent what "By" is referring to.

jianyangg commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,21 @@
+ * Represents a task that is ToDo in the Richie application
+ */
+
+public class Todo extends Task {
```
Consider using pascal case naming convention, from `Todo` to `ToDo` since "to-do" might not be considered as a single word (i.e., "todo").

jianyangg reviewed (commented):
LGTM! The code is generally written rather elegantly with good style and nice tweaks from the standard template.

Commit:
Rename produced jar file produced to duke-1.jar from duke.jar

Commit:
Abstract the conversion of a task to its data string to be stored in the data text file.

Storage saveCurrentTask() method gets the task description and boolean about whether it is done and converts them into data string to be stored in data text file.

There are multiple levels of abstraction in the saveCurrentTask() method  making it difficult to read. There is also a lot of code duplication.

Creating a new abstract method toDataString() into the Task class ensures that all instances of classes that inherit from Task class will have its own adoption of a toDataString() method.

Let's create an abstract method in Task class called toDataString() and add the toDataString() method adaptation into all classes that inherit from Task, such as Deadline, Todo and Event.

Let's reduce the code duplication in the saveCurrentTask() method in the Storage class and remove the if else statements that differentiate the different types of tasks using instanceof operator.

Commit:
Improve readability of all methods in the Parser class.

1. convertInputDateAndTimeIntoLocalDateTime()

Year is represented by Integer.parseInt(dayMonthYearArray[2]) without any indication that it is equal to year. Same with month, day, hours and minutes.

convertInputDataAndTimeIntoLocalDateTime() method is difficult to read because of Integer.parseInt(dayMonthYearArray[i]). It is not clear what each element in dayMonthYearArray refers to immediately. Same with month, day, hours and minutes.

Make the code look simpler by assigning a variable year, month, day etc. to Integer.parseInt(dayMonthYearArray[i]).

2. Extra level of nesting due to  String[] stringArray = userInput.split(" ", 2); being out of each possible command else if section.

Leads to worse readability due to greater amount unnecessary nesting, moreover the line String[] stringArray = userInput.split(" ", 2); is not clear what it does.

Add a comment line above the line to explain what it does and also put it into the else if sections for commands that need to use the line so as to remove one layer of nesting.

Commit:
Improve readability of Richie and TaskList methods.

Confusing lines of code.

Confuses the reader what those lines of code mean and do

Add comments above lines of confusing code so as to help readers understand the flow better

Commit:
Add assertions in convertInputDateAndTimeIntoLocalDateTime() method in Parser class.

User input day, month, year, hours and minutes are directly fed into the LocalDate.of() method.

The inputs go to the .of() method without being checked whether they are valid time inputs which might throw DateTimeException which gives a generic error message stating that there is a problem with the date time format in general.

Add assertions to check whether the time inputs are valid being passed into the .of() method.

This can give the user more specific error messages regarding their inputs which will be more useful to the user in correcting their user input.

ricketytoc committed:
Merge pull request #2 from ricketytoc/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'master' of https://github.com/ricketytoc/ip

* 'master' of https://github.com/ricketytoc/ip:
  Improve readability of Richie and TaskList methods.
  Improve readability of all methods in the Parser class.
  Abstract the conversion of a task to its data string to be stored in the data text file.

Commit:
Merge branch 'master' into branch-A-Assertions

* master:
  Improve readability of Richie and TaskList methods.
  Improve readability of all methods in the Parser class.
  Abstract the conversion of a task to its data string to be stored in the data text file.

# Conflicts:
#	src/main/java/duke/Parser.java

ricketytoc committed:
Merge pull request #1 from ricketytoc/branch-A-Assertions

Add Assertions

Commit:
Merge branch 'master' of https://github.com/ricketytoc/ip

* 'master' of https://github.com/ricketytoc/ip:
  Add assertions in convertInputDateAndTimeIntoLocalDateTime() method in Parser class.

Commit:
Increase flexibility of the find command.

find command only returns tasks with descriptions that match the keyword exactly.

This is not flexible as user might want to find a range of tasks that share similar letters.

Prioritises the tasks with descriptions that match more letters of the keyword always starting from the keyword first letter. Loop through the keyword substrings until left with only the first letter of the keyword. If the task description contains the substring of the keyword, then add the task into the result. Use a HashMap to track which tasks have already been added and which have not been added to the result array. If the task description contains the keyword entirely it will definitely be shown in the result, but if the description only contains a substring that is a part of the keyword, then it might not be added to the result if the limit has been reached. The limit can be set in the FindCommand.

Gives the user more flexibility when searching for a task.

Commit:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Increase flexibility of the find command.

Commit:
Improve readability of the filterTaskByKeyword method in TaskList

Commit:
Add comment to improve readability of filterTaskByKeyword method in TaskList class

Commit:
Improved the code quality in the Parser class by splitting the parse method into multiple smaller methods

Commit:
Fix bug where program will not reach the last else statement in the Parser parse method

Commit:
improve the code quality

Commit:
Improved code quality of the Parser and TaskList Class

Commit:
Add Ui.png in Docs folder

ricketytoc committed:
Update README.md

Commit:
Merge branch 'master' of https://github.com/ricketytoc/ip

* 'master' of https://github.com/ricketytoc/ip:
  Update README.md

Commit:
Add a new data entry in the data file

ricketytoc committed:
Update README.md

Commit:
Merge branch 'master' of https://github.com/ricketytoc/ip

* 'master' of https://github.com/ricketytoc/ip:
  Update README.md

Commit:
Allow for storing of tasks in the data file
